admin/tests/test_integration.py

```diff
@@ -430,3 +433,100 @@ def test_sdconfig_enable_https_on_source_interface():
+        fetch = +refs/heads/*:refs/remotes/origin/*
+'''
+
+
+@pytest.fixture
+def securedrop_git_repo(tmpdir):
+    os.chdir(str(tmpdir))
+    # Clone the SecureDrop repository into the temp directory.
+    cmd = ['git', 'clone',
+           'https://github.com/freedomofpress/securedrop.git']
+    subprocess.check_call(cmd)
+    os.chdir(os.path.join(str(tmpdir), 'securedrop/admin'))
+    subprocess.check_call('git reset --hard'.split())
+    # Now we will put in our own git configuration
+    with io.open('../.git/config', 'w') as fobj:
+        fobj.write(GIT_CONFIG)
+    # Let us move to an older tag
+    subprocess.check_call('git checkout 0.6'.split())
+    yield tmpdir
+
+
+# This class is to test all the git related operations.
+class TestGitOperations:
+    def test_check_for_update_when_updates_needed(self, securedrop_git_repo):
+        cmd = os.path.join(os.path.dirname(CURRENT_DIR),
+                           'securedrop_admin/__init__.py')
+        ansible_base = os.path.join(str(securedrop_git_repo),
+                                    'securedrop/install_files/ansible-base')
+        fullcmd = 'python {0} --root {1} check_for_updates'.format(
+                  cmd, ansible_base)
+        child = pexpect.spawn(fullcmd)
+        child.expect('Update needed', timeout=20)
+
+        child.expect(pexpect.EOF, timeout=10)  # Wait for CLI to exit
+        child.close()
+        assert child.exitstatus == 0
+        assert child.signalstatus is None
+
+    def test_check_for_update_when_updates_not_needed(self,
+                                                      securedrop_git_repo):
+        # Determine latest production tag using GitHub release object
+        github_url = 'https://api.github.com/repos/freedomofpress/securedrop/releases/latest'  # noqa: E501
+        latest_release = requests.get(github_url).json()
+        latest_tag = str(latest_release["tag_name"])
+
+        subprocess.check_call(["git", "checkout", latest_tag])
+
+        cmd = os.path.join(os.path.dirname(CURRENT_DIR),
+                           'securedrop_admin/__init__.py')
+        ansible_base = os.path.join(str(securedrop_git_repo),
+                                    'securedrop/install_files/ansible-base')
+        fullcmd = 'python {0} --root {1} check_for_updates'.format(
+                  cmd, ansible_base)
+        child = pexpect.spawn(fullcmd)
+        child.expect('All updates applied', timeout=20)
+
+        child.expect(pexpect.EOF, timeout=10)  # Wait for CLI to exit
+        child.close()
+        assert child.exitstatus == 0
+        assert child.signalstatus is None
+
+    def test_update(self, securedrop_git_repo):
+        gpgdir = os.path.join(os.path.expanduser('~'), '.gnupg')
+        os.mkdir(gpgdir)
+
+        # Set a reliable default keyserver for the tests.
+        with open(os.path.join(gpgdir, 'gpg.conf'), 'a') as f:
+            f.write('keyserver hkp://pgp.mit.edu')
```

Comment:
Non blocking remark: If run locally this will modify the `gpg.conf` of the user running the test and accumulate which is not harmful but not great either. However, as a developer I run tests using `make -C admin test` which protects my `~/.gnupg` directory because it is not exposed in the container.